File: [23.12]Design-Pattern/Java/src/behavioral/state/OffState.java
package behavioral.state;

public class OffState implements PowerState {
    private OffState() {
    }

    private static class SingleInstanceHolder {
        private static final OffState INSTANCE = new OffState();
    }

    public static OffState getInstance() {
        return OffState.SingleInstanceHolder.INSTANCE;
    }

    @Override
    public void powerButtonPush(Laptop laptop) {
        System.out.println("노트북 전원 ON");
        laptop.setPowerState(OnState.getInstance());
    }

    @Override
    public void typeButtonPush() {
        System.out.println("무반응");
    }

    @Override
    public String toString() {
        return "전원 상태 OFF";
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/behavioral/state/OffState.kt
package behavioral.state

object OffState : State {
    override fun powerButtonPush(laptop: Laptop) {
        println("OFF -> ON")
        laptop.setState(OnState)
    }

    override fun typeButtonPush() {
        println("무반응")
    }

    override fun toString(): String {
        return "현재 상태 OFF"
    }


}